src/background/background.ts

```diff
@@ -17,12 +17,36 @@ import { ExternalMethods, MessageFromContentScript } from '@content-scripts/mess
 // Listen for install event
 chrome.runtime.onInstalled.addListener(async details => {
   if (details.reason === 'install' && !IS_TEST_ENV) {
+    // The app url is saved for use after onboarding
+    const queryString = window.location.search;
+    const urlParams = new URLSearchParams(queryString);
+    const appUrl = urlParams.get('app');
+    if (appUrl) window.localStorage.setItem('appUrl', appUrl);
+
     await chrome.tabs.create({
       url: chrome.runtime.getURL(`full-page.html#${ScreenPaths.INSTALLED}`),
     });
   }
 });
 
+chrome.runtime.onMessage.addListener(message => {
+  // When onboarding is complete, redirect/reload the app tab
+  if (message === 'onboarding-complete') {
+    const appUrl = window.localStorage.getItem('appUrl') || undefined;
+    chrome.tabs.query({ lastFocusedWindow: true }, async tabs => {
+      const appTab = tabs?.find(tab => tab.url === appUrl);
+      if (appTab?.id) {
+        chrome.tabs.update(appTab?.id, { active: true });
+        chrome.tabs.reload(appTab?.id);
+      } else {
+        await chrome.tabs.create({
+          url: appUrl,
+        });
+      }
+    });
+  }
+});
```

Comment:
There's already a `chrome.runtime.onMessage` listener on line 89. Maybe better extend that one than create a second?

We'll need to consider how to add functionality to these listeners, but for now just breaking them up into separate functions would be a big improvement to readability:

```ts
function redirectAppTab(msg: Type) {...}

function vaultMessageHandler() {...}

chrome.runtime.onMessage.addListener((message: VaultActions, sender, sendResponse) => {
  if (message === 'onboarding-complete') redirectAppTab(message);

  void vaultMessageHandler(message, sender, sendResponse);

  // Specify that we are responding async
  return true;
});
```

This is where Rxjs can help out, but that's for another time

Reply:
Yeah, I debated about this bc how it is set up is so specific and hard to add to atm. What you suggested makes sense for now though! Will do.